src/core/analysis/mcmc/Mcmcmc.cpp

```diff
@@ -239,9 +239,48 @@ void Mcmcmc::checkpoint( void ) const
     for (size_t i = 0; i < num_chains; ++i)
     {
         
-        if ( chains[i] != NULL )
+        if ( chains[ chainForHeatIndex(i) ] != NULL )
         {
-            chains[i]->checkpoint();
+            // get the preliminary checkpoint file
+            path f = chains[ chainForHeatIndex(i) ]->getCheckpointFile();
+            
+            std::string path_string = f.string();
+            std::string test_string = "_chain_";
+            path chain_file_name;
+            
```

Comment:
I wonder how robust is the chain index replacing mechanism to the initial user specified checkpoint file name (i.e., the value provided to the `checkpointFile` argument of the `run` function). For example, if checkpointFile contains '_chain_' or multiple "."s (or no extension at all), would the search-and-replace way of specifying the dynamic chain-specific checkpoint file name still work as intended? I know those are all edge cases, but as we have no control on the user-specified `checkpointFile` argument, it would be nice if the implementation could make fewer assumptions. It seems that things would be easier if we could somehow just store the base file name (without the chain index part appended) and then append that part when looping over the chains here.

Reply:
You are totally right on all counts. I've given this some further thought and found an easy way to circumvent the problem, which gets rid of the whole index replacing mechanism and works with arbitrary file names. We simply use the `Mcmcmc::setCheckpointFile()` to set a `base_checkpoint_file_name`, generate chain-specific file names by appending to it, and then use `Mcmc::setCheckpointFile()` to set the checkpoint file names of the individual chains to what we have generated. It works just as well, but it's much more robust, and makes for cleaner and more concise code.